compiler/dcalc/optimizations.ml

```diff
@@ -248,9 +248,8 @@ let program_map
     (fun new_scopes -> { p with scopes = new_scopes })
     (scopes_map t ctx p.scopes)
 
-let optimize_program (p : 'm program) : untyped program =
+let optimize_program (p : 'm program) : 'm program =
   Bindlib.unbox
     (program_map partial_evaluation
        { var_values = Var.Map.empty; decl_ctx = p.decl_ctx }
        p)
-  |> Program.untype
```

Comment:
This one, yes: it only does reductions so that is easy to check. The one in `lcalc`, which I haven't merged yet with this one, I haven't checked yet, but it seems straightforward as well.